File: hellojava/src/test/java/com/fathzer/chess/PieceTest.java
package com.fathzer.chess;

import static org.junit.jupiter.api.Assertions.*;

import java.util.stream.Stream;

import static com.fathzer.chess.Piece.*;

import org.junit.jupiter.api.Test;

class PieceTest {

    @Test
    void testFromCode() {
        // Test all valid piece codes
        assertEquals(WHITE_PAWN, fromCode('P'));
        assertEquals(WHITE_KNIGHT, fromCode('N'));
        assertEquals(WHITE_BISHOP, fromCode('B'));
        assertEquals(WHITE_ROOK, fromCode('R'));
        assertEquals(WHITE_QUEEN, fromCode('Q'));
        assertEquals(WHITE_KING, fromCode('K'));
        assertEquals(BLACK_PAWN, fromCode('p'));
        assertEquals(BLACK_KNIGHT, fromCode('n'));
        assertEquals(BLACK_BISHOP, fromCode('b'));
        assertEquals(BLACK_ROOK, fromCode('r'));
        assertEquals(BLACK_QUEEN, fromCode('q'));
        assertEquals(BLACK_KING, fromCode('k'));
        
        // Test invalid code
        assertNull(fromCode('X'));
        assertNull(fromCode(' '));
    }

    @Test
    void testGetCode() {
        // Test a sample of pieces
        assertEquals('P', WHITE_PAWN.getCode());
        assertEquals('N', WHITE_KNIGHT.getCode());
        assertEquals('p', BLACK_PAWN.getCode());
        assertEquals('k', BLACK_KING.getCode());
    }

    @Test
    void testIsWhite() {
        // Test white pieces
        Stream.of(WHITE_PAWN, WHITE_KNIGHT, WHITE_BISHOP, WHITE_ROOK, WHITE_QUEEN, WHITE_KING)
            .forEach(piece -> assertTrue(piece.isWhite()));
        
        // Test black pieces
        Stream.of(BLACK_PAWN, BLACK_KNIGHT, BLACK_BISHOP, BLACK_ROOK, BLACK_QUEEN, BLACK_KING)
            .forEach(piece -> assertFalse(piece.isWhite()));
    }

    @Test
    void testCanBeCapturedBy() {
        // Test white pieces can be captured by black and not by white
        Stream.of(WHITE_PAWN, WHITE_KNIGHT, WHITE_BISHOP, WHITE_ROOK, WHITE_QUEEN, WHITE_KING)
            .forEach(piece -> {
                assertTrue(piece.canBeCapturedBy(false));
                assertFalse(piece.canBeCapturedBy(true));
            });
        // Test black pieces can be captured by white and not by black
        Stream.of(BLACK_PAWN, BLACK_KNIGHT, BLACK_BISHOP, BLACK_ROOK, BLACK_QUEEN, BLACK_KING)
            .forEach(piece -> {
                assertTrue(piece.canBeCapturedBy(true));
                assertFalse(piece.canBeCapturedBy(false));
            });        
   }

    @Test
    void testBlocker() {
        // Test BLOCKER specific properties
        assertFalse(BLOCKER.canBeCapturedBy(true));
        assertFalse(BLOCKER.canBeCapturedBy(false));
    }
}

File: hellogo/pkg/chess/piece_test.go
package chess

import (
	"testing"
)

func TestFromCode(t *testing.T) {
	// Test all valid piece codes
	tests := []struct {
		code rune
		want Piece
	}{
		{'P', WHITE_PAWN},
		{'N', WHITE_KNIGHT},
		{'B', WHITE_BISHOP},
		{'R', WHITE_ROOK},
		{'Q', WHITE_QUEEN},
		{'K', WHITE_KING},
		{'p', BLACK_PAWN},
		{'n', BLACK_KNIGHT},
		{'b', BLACK_BISHOP},
		{'r', BLACK_ROOK},
		{'q', BLACK_QUEEN},
		{'k', BLACK_KING},
	}
	for _, tt := range tests {
		got := FromCode(tt.code)
		if got == nil || *got != tt.want {
			t.Errorf("FromCode(%q) = %v, want %v", tt.code, got, tt.want)
		}
	}
	// Test invalid code
	if got := FromCode('X'); got != nil {
		t.Errorf("FromCode('X') = %v, want nil", got)
	}
	if got := FromCode(' '); got != nil {
		t.Errorf("FromCode(' ') = %v, want nil", got)
	}
}

func TestGetCode(t *testing.T) {
	if WHITE_PAWN.GetCode() != 'P' {
		t.Errorf("WHITE_PAWN.GetCode() = %c, want 'P'", WHITE_PAWN.GetCode())
	}
	if WHITE_KNIGHT.GetCode() != 'N' {
		t.Errorf("WHITE_KNIGHT.GetCode() = %c, want 'N'", WHITE_KNIGHT.GetCode())
	}
	if BLACK_PAWN.GetCode() != 'p' {
		t.Errorf("BLACK_PAWN.GetCode() = %c, want 'p'", BLACK_PAWN.GetCode())
	}
	if BLACK_KING.GetCode() != 'k' {
		t.Errorf("BLACK_KING.GetCode() = %c, want 'k'", BLACK_KING.GetCode())
	}
}

func TestIsWhitePiece(t *testing.T) {
	// Test white pieces
	whitePieces := []*Piece{&WHITE_PAWN, &WHITE_KNIGHT, &WHITE_BISHOP, &WHITE_ROOK, &WHITE_QUEEN, &WHITE_KING}
	for _, p := range whitePieces {
		if !p.IsWhitePiece() {
			t.Errorf("%v.IsWhitePiece() = false, want true", p)
		}
	}
	// Test black pieces
	blackPieces := []*Piece{&BLACK_PAWN, &BLACK_KNIGHT, &BLACK_BISHOP, &BLACK_ROOK, &BLACK_QUEEN, &BLACK_KING}
	for _, p := range blackPieces {
		if p.IsWhitePiece() {
			t.Errorf("%v.IsWhitePiece() = true, want false", p)
		}
	}
}

func TestCanBeCapturedBy(t *testing.T) {
	// Test white pieces can be captured by black
	whitePieces := []*Piece{&WHITE_PAWN, &WHITE_KNIGHT, &WHITE_BISHOP, &WHITE_ROOK, &WHITE_QUEEN, &WHITE_KING}
	for _, p := range whitePieces {
		if !p.CanBeCapturedBy(false) {
			t.Errorf("%v.CanBeCapturedBy(false) = false, want true", p)
		}
		if p.CanBeCapturedBy(true) {
			t.Errorf("%v.CanBeCapturedBy(true) = true, want false", p)
		}
	}
	// Test black pieces can be captured by white
	blackPieces := []*Piece{&BLACK_PAWN, &BLACK_KNIGHT, &BLACK_BISHOP, &BLACK_ROOK, &BLACK_QUEEN, &BLACK_KING}
	for _, p := range blackPieces {
		if !p.CanBeCapturedBy(true) {
			t.Errorf("%v.CanBeCapturedBy(true) = false, want true", p)
		}
		if p.CanBeCapturedBy(false) {
			t.Errorf("%v.CanBeCapturedBy(false) = true, want false", p)
		}
	}
}

func TestBlocker(t *testing.T) {
	if (&BLOCKER).CanBeCapturedBy(true) {
		t.Errorf("BLOCKER.CanBeCapturedBy(true) = true, want false")
	}
	if (&BLOCKER).CanBeCapturedBy(false) {
		t.Errorf("BLOCKER.CanBeCapturedBy(false) = true, want false")
	}
}
